kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/metrics/MetricsReportTestUtils.scala

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (2024) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.defaults.metrics
+
+import java.util
+
+import scala.collection.mutable.ArrayBuffer
+
+import io.delta.kernel.defaults.utils.TestUtils
+import io.delta.kernel.engine._
+import io.delta.kernel.metrics.MetricsReport
+
+/**
+ * Test utilities for testing the Kernel-API created [[MetricsReports]]s.
+ *
+ * We test [[MetricsReport]]s in the defaults package so we can use real tables and avoid having
+ * to mock both file listings AND file contents.
+ */
+trait MetricsReportTestUtils extends TestUtils {
+
+  // For now this just uses the default engine since we have no need to override it, if we would
+  // like to use a specific engine in the future for other tests we can simply add another arg here
+  /**
+   * Executes [[f]] using a special engine implementation to collect and return metrics reports.
+   * If [[expectException]], catches any exception thrown by [[f]] and returns it with the reports.
+   */
+  def collectMetricsReports(
+    f: Engine => Unit, expectException: Boolean): (Seq[MetricsReport], Option[Exception]) = {
+    // Initialize a buffer for any metric reports and wrap the engine so that they are recorded
+    val reports = ArrayBuffer.empty[MetricsReport]
+    if (expectException) {
+      val e = intercept[Exception](
+        f(new EngineWithInMemoryMetricsReporter(reports, defaultEngine))
+      )
+      (reports, Some(e))
+    } else {
+      f(new EngineWithInMemoryMetricsReporter(reports, defaultEngine))
+      (reports, Option.empty)
+    }
+  }
+
+  def resolvePath(path: String): String = {
+    defaultEngine.getFileSystemClient.resolvePath(path)
+  }
+
+  /**
+   * Wraps an {@link Engine} to implement the metrics reporter such that it appends any reports
+   * to the provided in memory buffer.
+   */
+  class EngineWithInMemoryMetricsReporter(buf: ArrayBuffer[MetricsReport], baseEngine: Engine)
```

Comment:
why does it take in the buffer to which we want to append the metrics?

would it ever take in a buffer that is non-empty? all the usages I see use `val reports = ArrayBuffer.empty[MetricsReport]`

do our engine APIs not make it simple enough to get the metrics reports?

Reply:
I guess this implementation could create the empty buffer itself, but we would still need it to be externally accessible so we can access the reports. I don't think there's much value added by doing this and I think it would actually make `collectMetricsReports` a bit more complex.

> do our engine APIs not make it simple enough to get the metrics reports?

I think this is pretty simple :) we don't provide any out-of-the-box functionality to buffer the reports in memory thus we need to implement an "in memory metrics reporter". We add this to the existing defaultEngine reporter such that we have both this one, and the default logging reporter so we do log the reports in the test suite as well.

Reply:
Sounds good to me. Thanks for explaining